Handle extensible types

zoggy opened the pull request.

zoggy commented:
Compilation should be ok now that changes in ppx_deriving has been merged.

whitequark commented on src/ppx_deriving_yojson.ml:
```diff
@@ -210,147 +210,430 @@ let wrap_runtime decls =
+        match type_decl.ptype_manifest with
+        | Some ({ ptyp_desc = Ptyp_constr ({ txt = lid }, args) } as manifest) ->
+            let ser = ser_expr_of_typ manifest in
+            let lid = Ppx_deriving.mangle_lid (`PrefixSuffix ("M_", "to_yojson")) lid in
```
Why the trailing underscore?

zoggy replied:
Because mangle_lid will add one, and I usually generate ids with two underscores to prevent conflicts (I agree that in this case the probability of a conflict is rather low. Same for t__ . Do you prefer only one ?

whitequark replied:
Yeah, I think one is fine.

whitequark commented on src/ppx_deriving_yojson.ml:
```diff
@@ -210,147 +210,430 @@ let wrap_runtime decls =
+            in
+            let mod_name = "M__"^to_yojson_name in
+            let typ = Type.mk ~kind: (Ptype_record
+               [ Type.field ~mut: Mutable (mknoloc "f") ty ]) (mknoloc "t__to_yojson")
```
Ditto about underscores. Also, I'm not sure about hardcoded `t_`...

zoggy replied:
What does bother you with hardcoded t_... ?

whitequark replied:
Oh, I realized why it is there. Still, there should only be one underscore.

whitequark commented on src/ppx_deriving_yojson.ml:
```diff
@@ -210,147 +210,430 @@ let wrap_runtime decls =
+              (fun var -> [%type: [%t var] -> Yojson.Safe.json]) type_decl
+            in
+            let ty = Typ.poly poly_vars (polymorphize_ser [%type: [%t typ] -> Yojson.Safe.json]) in
+            let default_fun = [%expr fun _ -> assert false] in
```
The error should be more informative than an assertion failure in literally nowhere (the location isn't recorded in `[%expr]`).

zoggy replied:
Ok, I will improve this.

whitequark commented:
OK, this isn't bad. I don't like the naming of externally visible modules (`M__`...) and that unit results are stored under the names of `_extend_of_yojson` and `_extend_to_yojson`.

whitequark commented on src_test/test_ppx_yojson.ml:
```diff
@@ -220,6 +220,32 @@ let test_nostrict ctxt =
+  struct
+    type 'a opentype = .. [@@deriving yojson]
+    type 'a opentype += A of 'a | B of string list [@@deriving yojson]
+  end
```
Can you also add a signature to test that the signatures for externally visible modules are derived correctly?

zoggy replied:
Good idea.

zoggy commented:
> ... unit results are stored under the names of _extend_of_yojson and _extend_to_yojson.

Since they end in a let rec, I can't use ``_`` nor ``()``. I added a type annotation in the form ``(_extend_to_yojson : unit)`` to ensure the resulting type.

I changed to use only one underscore rather than two in the generated module and type names. I also handled constructor rebind.

whitequark commented:
Well, they don't have to be in a let rec. You can return a non-`Recursive` structure item for these generated functions.

whitequark commented:
I've looked at everything else and it's good. I'll merge (and probably release) after you fix the `let rec`.

zoggy commented:
This is done.

whitequark commented:
How can I attribute you in a email announcement?

zoggy commented:
Zoggy's fine :)
